src/main/java/cloud/orbit/spring/actuate/ActorInfoContributorLifetimeExtension.java

```diff
@@ -0,0 +1,165 @@
+ DISCLAIMED. IN NO EVENT SHALL ELECTRONIC ARTS OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package cloud.orbit.spring.actuate;
+
+import org.springframework.boot.actuate.info.Info;
+import org.springframework.boot.actuate.info.InfoContributor;
+
+import cloud.orbit.actors.Actor;
+import cloud.orbit.actors.extensions.LifetimeExtension;
+import cloud.orbit.actors.runtime.AbstractActor;
+import cloud.orbit.concurrent.Task;
+
+import java.lang.ref.WeakReference;
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class ActorInfoContributorLifetimeExtension implements LifetimeExtension, InfoContributor
+{
+    private final Set<ActorInfoContributorReference> actorInfoContributorReferences = new HashSet<>();
+    private final OrbitRemoteReferenceService orbitRemoteReferenceService;
+
+    ActorInfoContributorLifetimeExtension(final OrbitRemoteReferenceService orbitRemoteReferenceService)
+    {
+        this.orbitRemoteReferenceService = orbitRemoteReferenceService;
+    }
+
+    @Override
+    public Task<?> preActivation(final AbstractActor<?> actor)
+    {
+        return doIfInfoContributor(actor, actorInfoContributorReferences::add);
+    }
+
+    @Override
+    public Task<?> preDeactivation(final AbstractActor<?> actor)
+    {
+        return doIfInfoContributor(actor, actorInfoContributorReferences::remove);
+    }
+
+    private Task<?> doIfInfoContributor(final AbstractActor<?> actor,
+                                        final Consumer<ActorInfoContributorReference> synchronizedAction)
+    {
+        if (actor instanceof InfoContributor)
+        {
+            ActorInfoContributorReference actorInfoContributorReference = new ActorInfoContributorReference(actor);
+            synchronized (actorInfoContributorReferences)
+            {
+                synchronizedAction.accept(actorInfoContributorReference);
+            }
+        }
+        return Task.done();
+    }
+
+    @Override
+    public void contribute(final Info.Builder builder)
+    {
+        createActorInfoObject().ifPresent(actorInfoObject -> builder.withDetail("actors", actorInfoObject));
+    }
+
+    private Optional<?> createActorInfoObject()
+    {
+        Set<ActorInfoContributorReference> actorInfoContributorReferencesCopy;
+        synchronized (actorInfoContributorReferences)
+        {
+            actorInfoContributorReferencesCopy = actorInfoContributorReferences.stream()
+                    .map(ActorInfoContributorReference::new)
```

Comment:
I think the code in this method can be optimized to use less memory, not lock, and do less iterations:

1. Remove the expired references in this iteration
1. You don't need to allocate a new ActorInfoContributorReference for the existing objects, just reuse the existing ones because they're immutable
1. Instead of the `removeAll` call later, simply replace that collection reference with the collection that you built here
1. Don't worry about synchronization at all because without removeAll, the collections are immutable. In the worst case you end up with some expired references that will be removed by a future call